Add starting_up_state parameter to Updater

redvinaa opened the pull request:
In the current implementation, when the Updater object is constructed, it sends off an "OK" signal. This could be problematic if we consider "Everything is *running* as expected" (definition of "OK" status as per REP107), that running is already initialized.

The default behavior doesn't change with this PR, so it wouldn't break anything.
Also, in the python implementation, None can be passed to skip "Node starting up" status publishing altogether.

redvinaa committed:
Add starting_up_state parameter to Updater

redvinaa committed:
Add starting_up_state parameter to Updater

redvinaa committed:
Lint

redvinaa committed:
Lint

redvinaa commented:
Hi! Can I get an update on this one?

If it wasn't merged because of this test case failing, can I get some help understanding the error? I'm sure a few minutes of someone who already understands these template-based tests can save me hours of trial and error.
After that I can also rebase to fix the conflict.

ct2034 commented:
Hi @redvinaa.
Could we please have a quick discussion about this? I am not sure if it is worth the extra overhead. Yes, 
> "Everything is running as expected" (definition of "OK" status as per REP107)

On the other hand, I also would not like my system starting in an error state. Your issue could also be solved by additional state, such as proposed in https://github.com/ros2/common_interfaces/issues/268. What is your opinion about that?

redvinaa commented:
Hi! Sorry for the long delay. It's possible that it would make sense to add new states, but that's a huge change and it appears to be moving very slowly. This PR on the other hand is not breaking, and I think is still useful until the new states get added.

ct2034 commented on diagnostic_updater/src/diagnostic_updater.cpp:
```diff
@@ -190,9 +193,13 @@ void Updater::publish(std::vector<diagnostic_msgs::msg::DiagnosticStatus> & stat
 {
+  if (starting_up_status_ < 0) {
+    return;
+  }
```
what is the purpose of this line?

redvinaa replied:
I don't remember anymore, removing

ct2034 commented on diagnostic_updater/include/diagnostic_updater/diagnostic_updater.hpp:
```diff
@@ -363,19 +363,23 @@ class Updater : public DiagnosticTaskVector
-  explicit Updater(NodeT node, double period = 1.0)
+  explicit Updater(
+    NodeT node, double period = 1.0,
+    int starting_up_status = diagnostic_msgs::msg::DiagnosticStatus::OK)
```
```suggestion
    unsigned char starting_up_status = diagnostic_msgs::msg::DiagnosticStatus::OK)
```

ct2034 commented on diagnostic_updater/include/diagnostic_updater/diagnostic_updater.hpp:
```diff
@@ -385,7 +389,8 @@ class Updater : public DiagnosticTaskVector
     std::shared_ptr<rclcpp::node_interfaces::NodeTopicsInterface> topics_interface,
-    double period = 1.0);
+    double period = 1.0,
+    int starting_up_status = diagnostic_msgs::msg::DiagnosticStatus::OK);
```
```suggestion
    unsigned char starting_up_status = diagnostic_msgs::msg::DiagnosticStatus::OK);
```

ct2034 commented on diagnostic_updater/include/diagnostic_updater/diagnostic_updater.hpp:
```diff
@@ -478,6 +483,7 @@ class Updater : public DiagnosticTaskVector
   rclcpp::Publisher<diagnostic_msgs::msg::DiagnosticArray>::SharedPtr publisher_;
   rclcpp::Logger logger_;
 
+  int starting_up_status_;
```
```suggestion
  unsigned char starting_up_status_;
```

ct2034 commented on diagnostic_updater/src/diagnostic_updater.cpp:
```diff
@@ -44,7 +44,9 @@ Updater::Updater(
-  std::shared_ptr<rclcpp::node_interfaces::NodeTopicsInterface> topics_interface, double period)
+  std::shared_ptr<rclcpp::node_interfaces::NodeTopicsInterface> topics_interface,
+  double period,
+  int starting_up_status)
```
```suggestion
  unsigned char starting_up_status)
```

redvinaa committed:
Change to unsigned char

redvinaa committed:
Remove condition

redvinaa commented:
I think the xmllint failure is not coming from this PR (no xml was modified). Otherwise, all requested changes have been addressed.

Noel215 committed:
Fix example relative path (#550)

Noel215 committed:
Implement onParametersSet for handling only analyzers node parameters (#551)

The previous implementation is using a callback that is invoked for all
the parameter events of all ROS 2 nodes and it filters by node name.
It could lead to race conditions when setting parameters depending on
the depth size of the callback.

With this new approach, the callback is only triggered when setting
parameters for the analyzers node, avoiding possible race conditions
when setting the parameters.

Co-authored-by: Christian Henkel <christian.henkel2@de.bosch.com>

mini-1235 committed:
Get rid of deprecated rclcpp::spin_some() (#563)

Signed-off-by: mini-1235 <mauricepurnawan@gmail.com>

ct2034 committed:
-Wreorder

Signed-off-by: Christian Henkel <christian.henkel2@de.bosch.com>

ct2034 committed:
Merge branch 'ros2' into starting-up-state
